haskell/cc.bzl

```diff
@@ -35,9 +35,11 @@ def cc_headers(ctx):
     for dep in ctx.attr.deps if CcSkylarkApiProviderHacked in dep
   ])
 
-  dirs = set.to_list(set.from_list([hdr.dirname for hdr in hdrs]))
+  flags = set.to_list(set.from_list(
+      [f for dep in ctx.attr.deps if hasattr(dep, "cc")
+         for f in dep.cc.compile_flags]))
```

Comment:
Also note that the cc_library "defines" and "includes" fields already have strong warnings in its docs:
https://docs.bazel.build/versions/master/be/c-cpp.html
For example:
```Unlike copts, these flags are added for the target and every rule that depends on it! Be very careful, since this may have far-reaching effects. When in doubt, add "-D" (or /D on Windows) flags to copts instead.```
Which indicates that we shouldn't have to worry about conflicts.

Another option is to assemble the set of compile flags manually, basically replicating the logic in `compile_flags`.  For example, in `rules_go`:
https://github.com/bazelbuild/rules_go/blob/master/go/private/rules/cgo.bzl#L187
Compare:
https://github.com/bazelbuild/bazel/blob/fa135cf600a957cbcef0ca45f97d5a9009d40859/src/main/java/com/google/devtools/build/lib/rules/cpp/CcSkylarkApiProvider.java#L175